### PR TITLE
Test

### PR DESCRIPTION
**Pull request by gyc2432968**
merge to ultradian_test

### CODE REVIEW

File: mousestyles/ultradian/tests/test_ultradian.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+def seasonal_decomposition_test(strain=0, mouse=1, feature='AS', bin_width=30, period_length=10):
+    resul = seasonal_decomposition(strain, mouse, feature, bin_width, period_length)
+        assert tyep(result) == statsmodels.tsa.seasonal.DecomposeResult
+
+#strain_seasonal is to produce a plot, so I think we can check the plot directly istead of writing a test function
+
+#check the outcome range of mix_strain, the result is a p-value, so it's between 0 and 1
+def mix_strain_test(data, feature):
+    result = mix_strain(data, feature)
+        assert  0 < result < 1
```
Comment (boyinggong) [inline mousestyles/ultradian/tests/test_ultradian.py]: Could you please fix the style for pep8 here?

File: mousestyles/ultradian/tests/test_ultradian.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    result = aggregate_interval(strain, mouse, feature, bin_width)
+        assert type(result) is pd.core.series.Series
+#Check the length of the outcome, should be 17280/bin_width
+def aggregate_interval_test_length(strain=0, mouse=1, feature="AS", bin_width=30):
+    result = aggregate_interval(strain, mouse, feature, bin_width)
+        assert len(result) == 17280/bin_width
+
+#Check the outcome data type of function aggregate_movement()
+def aggregate_movement_test(strain=0, mouse=1, bin_width=30):
+    result = aggregate_movement(strain,mouse, bin_width)
```
Comment (boyinggong) [inline mousestyles/ultradian/tests/test_ultradian.py]: Could you please fix style and check it by run make style in bash in the mousestyles folder?
Or you could do it in sublime text, either way is fine

File: mousestyles/ultradian/tests/test_ultradian.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+#Check the values in feature columns, all the values should be greater than 0
+def aggregate_data_test_feature(feature="AS", bin_width=30):
+    result = aggregate_data(feature,bin_width)
+    test = np.array(result[feature])
+    assert test.min() >= 0
+
+#Test for seasonal decomposition
+def seasonal_decomposition_test(strain=0, mouse=1, feature='AS', bin_width=30, period_length=10):
+    resul = seasonal_decomposition(strain, mouse, feature, bin_width, period_length)
+        assert tyep(result) == statsmodels.tsa.seasonal.DecomposeResult
```
Comment (boyinggong) [inline mousestyles/ultradian/tests/test_ultradian.py]: Did you actually run these test? I think you have som typos here `tyep`

**Comment by boyinggong** [comment]
Could you please add some input test here? For example, to test the input feature is valid or not, or to test if the mouse and strain number is integer. etc?

**Comment by boyinggong** [comment]
@gyc2432968 Richie has update the input check, could you please update the test? Also could you please fix the above problems? If you have any questions, please let me know. Thanks.

**Comment by gyc2432968** [comment]
@boyinggong I have done all the input test and add some style check, pls review. thanks

**Comment by YZhouEntheos** [comment]
This is great! But I think there is one test missed for function `strain_seasonal`. Should we also test that? @gyc2432968 @boyinggong

**Comment by gyc2432968** [comment]
@YZhouEntheos that is a function creating a plot, I think we can look at the plot directly.

**Comment by YZhouEntheos** [comment]
@gyc2432968 I think that function returns two object, one is a bumpy array for each mouse in one specific strain, one is a plot. Should we test the numpy array it returns?

**Comment by gyc2432968** [comment]
@YZhouEntheos Let me try it, should be quick to fix.

**Comment by YZhouEntheos** [comment]
@boyinggong Also should we split that function into two functions and move the visualisation one to the visualization subpackage?

**Comment by boyinggong** [comment]
@richiefeng @pigriver123 @HaoLyuKrystal Please review this.

**Comment by boyinggong** [comment]
@YZhouEntheos Good idea. I will do that later.

**Comment by richiefeng** [comment]
I just wonder are those test functions able to be executed when building the package locally? If it has no problems, you can merge it.

**Comment by gyc2432968** [comment]
@boyinggong @YZhouEntheos  I just added the test for strain_seasonal

**Comment by gyc2432968** [comment]
@richiefeng I just ran them, they can be executed.

**Comment by richiefeng** [comment]
@gyc2432968 That's great! Have you tried building the package?

**Comment by gyc2432968** [comment]
@richiefeng Just ran it locally, did not try building the package, I can try it.

**Comment by richiefeng** [comment]
@gyc2432968 It is good to make sure the package can be built after inserting new test functions. Otherwise it would cause some troubles.

**Comment by YZhouEntheos** [comment]
@gyc2432968 Pls run `make test` to check

**Comment by richiefeng** [comment]
If the package building fails at sometime, create a new virtual environment in Conda and redo it.

File: mousestyles/ultradian/tests/test_ultradian.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+import numpy as np
+import pandas as pd
+import statsmodels.api as sm
+import statsmodels.formula.api as smf
+from scipy.stats import chi2
+import statsmodels
+
+import mousestyles.data as data
+
```
Comment (YZhouEntheos) [inline mousestyles/ultradian/tests/test_ultradian.py]: You should `import mousestyles.ultradian as ultradian` here

File: mousestyles/ultradian/tests/test_ultradian.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+# aggregate_interval test
+
+
+def test_aggregate_interval():
+    with pytest.raises(ValueError) as msg1:
+        aggregate_interval(-1, 1, "M_IS", 30)
```
Comment (YZhouEntheos) [inline mousestyles/ultradian/tests/test_ultradian.py]: And change this to `ultradian.aggregate_interval(-1, 1, "M_IS", 30)`

File: mousestyles/ultradian/tests/test_ultradian.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    assert msg3.value.args[
+        0] == 'Bin width (minutes) must be a non-negative integer below 1440'
+# Check the outcome data type of function aggregate_movement()
+    result = aggregate_movement(0, 1, 30)
+    assert type(result) is pd.core.series.Series
+
+
+# Test for aggregate_data
+def test_aggregate_data(feature="AS", bin_width=30):
+    result = aggregate_data(feature, bin_width)
```
Comment (YZhouEntheos) [inline mousestyles/ultradian/tests/test_ultradian.py]: Same with here, change to `ultradian.aggregate_data(feature, bin_width)`

File: mousestyles/ultradian/tests/test_ultradian.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    assert test_1.max() < 24
+# Check the strain
+    test_2 = np.array(result['strain'])
+    assert len(np.unique(test_2)) == 3
+
+
+# Test for seasonal decomposition
+def test_seasonal_decomposition(strain=0, mouse=1, feature='AS',
+                                bin_width=30, period_length=10):
+    result = seasonal_decomposition(strain,
```
Comment (YZhouEntheos) [inline mousestyles/ultradian/tests/test_ultradian.py]: Another example, `ultradian.seasonal_decomposition(strain,`

**Comment by HaoLyuKrystal** [comment]
@gyc2432968 When you run the test for the functions of models, do you get some warnings? like not converge or non integer input? If so can we get rid off these warnings?

Also for my function: mix_strain, besides the p value output I also print out some summaries. I was wondering will these affect the test?  Do I need to remove these print out?

**Comment by gyc2432968** [comment]
@HaoLyuKrystal For your function, since you add many print in the function, it's a little bit hard to test it.

**Comment by gyc2432968** [comment]
@HaoLyuKrystal But we can still pass the test, only issue is that it has some more information than other test functions

**Comment by HaoLyuKrystal** [comment]
@gyc2432968 I can make the print an option, then when you do the testing you can make it not print.

**Pull request merged**